### ISSUE TITLE
Email/set with long from/to/cc/bcc name with non-ascii chars generates invalid rfc822 message

### ISSUE THREAD

**Issue by robmueller** [cyrusimap/cyrus-imapd#2293]
This call works:

```
[ "Email/set", {
  accountId => q{test1@robmtest.vm},
  create => { k1 => {
    mailboxIds => { "1bfc127d-a1e6-44f6-8a4c-fa4c6bed6efb" => JSON::true },
    from => [ { email => q{test1@robmtest.vm}, name => q{} } ],
    to => [ {
      email => q{foo@bar.com},
      name => "asd \x{529b}\x{9928}\x{5fc5}  asd \x{30ec}\x{30f1}\x{30b9}"
    } ]
    } }
}, "c1" ]
```

```
MIME-Version: 1.0
From: test1@robmtest.vm
Date: Wed, 14 Mar 2018 01:59:15 +0000
To: =?UTF-8?Q?asd_=E5=8A=9B=E9=A4=A8=E5=BF=85__asd_=E3=83=AC=E3=83=B1=E3=82=B9?= <foo@bar.com>
Subject: 
Message-ID: <485c566f-7987-4460-b922-ef78fa1c0aca@slotv1t01>
User-Agent: 
```

Though why is User-Agent empty?

But this call with a slightly longer name is totally broken.

```
[ "Email/set", {
  accountId => q{test1@robmtest.vm},
  create => { k1 => {
    mailboxIds => { "1bfc127d-a1e6-44f6-8a4c-fa4c6bed6efb" => JSON::true },
    from => [ { email => q{test1@robmtest.vm}, name => q{} } ],
    to => [ {
      email => q{foo@bar.com},
      name => "asd \x{529b}\x{9928}\x{5fc5}  asd \x{30ec}\x{30f1}\x{30b9} asd  \x{3b1}\x{3bc}\x{3b5}\x{3c4}"
    } ]
    } }
}, "c1" ]
```

```
MIME-Version: 1.0
From: test1@robmtest.vm
Date: Wed, 14 Mar 2018 01:58:43 +0000
To: =?UTF-8?Q?=3D=3FUTF-8=3FQ=3Fasd=5F=3DE5=3D8A=3D9B=3DE9=3DA4=3DA8=3DE5=3D?=
 =?UTF-8?Q?BF=3D85=5F=5Fasd=5F=3DE3=3D83=3DAC=3DE3=3D83=3DB1=3DE3=3D82=3D?=
 =?UTF-8?Q?B9=3F=3D?=
 =?UTF-8?Q?=3D=3FUTF-8=3FQ=3F=5Fasd=5F=5F=3DCE=3DB1=3DCE=3DBC=3DCE=3DB5?=
 =?UTF-8?Q?=3DCF=3D84=3F=3D_<foo@bar.com>?=
Subject: 
Message-ID: <d3e76b7c-fbce-422b-b597-601dda9eef2d@slotv1t01>
User-Agent: 
```

Looks like a bunch of double encoding happening when folding happens.

**Comment by rsto** [comment]
Interesting! Can this wait until the new Email object type has been implemented for Email/set? It's basically a rewrite of all the Email/get and Email/set code.

**Comment by robmueller** [comment]
There's a separate, but possibly related issue. It's currently not wrapping long to/cc/bcc lines:

```
To: =?UTF-8?Q?Rob_M=C3=BCller_=E2=A6=BF?= <robm@example.fm>, =?UTF-8?Q?Rob_M=C3=BCller_=E2=A6=BF?= <robm@example.fm>, =?UTF-8?Q?Rob_M=C3=BCller_=E2=A6=BF?= <robm@example.fm>, =?UTF-8?Q?Rob_M=C3=BCller_=E2=A6=BF?= <robm@example.fm>, =?UTF-8?Q?Rob_M=C3=BCller_=E2=A6=BF?= <robm@example.fm>, =?UTF-8?Q?Rob_M=C3=BCller_=E2=A6=BF?= <robm@example.fm>
```

I presume the code that does the [ { email => ..., name => ... }, ... ] -> rfc822 email header conversion is relatively independent of this Email/set rewrite? I'd really like if you could fix this issue and the above wrapping issue first and commit to master, it'll get us really close to actually being able to use JMAP on FastMail beta for day-to-day use.

**Comment by rsto** [comment]
It's going to be a full-on rewrite.

But I'll see if I can fix that in the legacy Email implementation on master.

**Comment by rsto** [comment]
Fixed in 43dff5f081fcb93615283a17bdcf8d6547c1675d, a4c8b307e7f37042166003ab385632a04e3a70f6, bfe10d1f3b79eb022460f9f4a472a23caf09a193
